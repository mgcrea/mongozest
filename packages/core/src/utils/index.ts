export * from './logger';
export * from './objectId';
export * from './traversing';
