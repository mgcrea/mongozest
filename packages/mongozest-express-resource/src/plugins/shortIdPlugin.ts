import {Resource} from '..';

const SID_REGEX = /^[a-z0-9\-\_]{7,14}$/i;

export default function shortIdPlugin(resource: Resource, {sidKey = '_sid'} = {}) {
  resource.addFilterUrlParams({
    [sidKey]: SID_REGEX
  });
}
