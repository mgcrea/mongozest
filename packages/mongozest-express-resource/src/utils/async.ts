export const asyncHandler = (wrappedFunction: Function) =>
  function asyncWrap(...args: any[]) {
    const next = args[args.length - 1];
    return new Promise(function(resolve) {
      resolve(wrappedFunction(...args));
    }).catch(next);
  };
