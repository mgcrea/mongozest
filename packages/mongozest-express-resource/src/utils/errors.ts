import createError from 'http-errors';
import {Request, Response, NextFunction} from 'express';
import {MongoError} from '@mgcrea/mongozest';

export const mongoErrorMiddleware = (err: any, req: Request, res: Response, next: NextFunction) => {
  // Convert MongoError to HttpErrors
  if (err instanceof MongoError) {
    const {message, errmsg, code = 0} = err;
    const errorMessage = errmsg || message || 'Unknown Mongo Error';
    // Handle Validation errors
    if (code === 121) {
      next(createError(422, errorMessage));
      return;
    }
    // Handle duplicate key errors (check me!)
    if ([110, 11000].includes(code)) {
      next(createError(409, errorMessage));
      return;
    }
    next(createError(400, errorMessage));
    return;
  }
  next(err);
};
