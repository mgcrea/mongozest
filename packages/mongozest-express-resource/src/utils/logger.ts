import chalk from 'chalk';
import {inspect as defaultInspect} from 'util';

export const inspect = (maybeObject: object) =>
  defaultInspect(maybeObject, {compact: true, colors: true, depth: Infinity, breakLength: Infinity});

export const log = (maybeString: string) =>
  console.log(`${chalk.gray(new Date().toISOString())} - 🏢 resource: ${maybeString}`);
