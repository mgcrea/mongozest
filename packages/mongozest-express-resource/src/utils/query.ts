// import JSON5 from 'json5';
// import {defaults, pick, isPlainObject, mapValues} from 'lodash';
// import createError from 'http-errors';

// export const parseQueryOptions = (req, whitelist = []) => {
//   // Only pick valid query options
//   let queryOptions = defaults(pick(req.query, whitelist), {});
//   // Parse any JSON-like options
//   queryOptions = mapValues(queryOptions, (value, key) => {
//     if (value && typeof value === 'string' && (value[0] === '{' || value[0] === '[')) {
//       try {
//         return JSON5.parse(value);
//       } catch (err) {
//         throw createError(400, `Failed to parse query field=\`${key}\``);
//       }
//     }
//     return value;
//   });
//   // Extend with req parameters
//   if (req.params && Object.keys(req.params).length) {
//     queryOptions.query = Object.assign(queryOptions.query || {}, req.params);
//   }
//   return queryOptions;
// };

// export const parseQuerySelect = req => {
//   const query = parseQueryOptions(req, ['select']);
//   if (isPlainObject(query.select)) {
//     return query.select;
//   }
//   const querySelectAsArray = !Array.isArray(query.select) ? query.select.toString().split(' ') : query.select;
//   return querySelectAsArray.reduce((soFar, value) => {
//     if (value.startsWith('+')) {
//       soFar[value.substr(1)] = 1;
//     } else if (value.startsWith('-')) {
//       soFar[value.substr(1)] = 0;
//     } else {
//       soFar[value] = 1;
//     }
//     return soFar;
//   }, {});
// };

// export const applyQueryOptions = (query, queryOptions = {}) => {
//   if (queryOptions.distinct) {
//     query.distinct(queryOptions.distinct);
//   }
//   if (queryOptions.limit > 0) {
//     query.limit(queryOptions.limit * 1);
//   }
//   if (queryOptions.select) {
//     if (Array.isArray(queryOptions.select)) {
//       query.select(queryOptions.select.join(' '));
//     } else {
//       query.select(queryOptions.select);
//     }
//   }
//   if (queryOptions.populate) {
//     query.populate(queryOptions.populate);
//   }
//   if (queryOptions.sort) {
//     query.sort(queryOptions.sort);
//   }
//   return query;
// };
