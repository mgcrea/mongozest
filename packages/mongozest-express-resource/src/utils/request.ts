export const parseBodyAsUpdate = (body: {[s: string]: any}) =>
  Object.keys(body).reduce(
    (soFar: {[s: string]: any}, key: string) => {
      if (!key.startsWith('$')) {
        soFar.$set[key] = body[key];
      } else {
        soFar[key] = body[key];
      }
      return soFar;
    },
    {$set: {}}
  );

export const asyncHandler = (wrappedFunction: Function) =>
  function asyncWrap(...args: any[]) {
    const next = args[args.length - 1];
    return new Promise(function(resolve) {
      resolve(wrappedFunction(...args));
    }).catch(err => {
      d(err);
      next(err);
    });
  };
