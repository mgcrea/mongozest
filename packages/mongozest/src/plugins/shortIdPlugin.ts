import shortid from 'shortid';
import {memoize} from 'lodash';
// @types
import {Model} from '..';
import {FindOneOptions, UpdateQuery, UpdateWriteOpResult, ReplaceOneOptions, ObjectId} from 'mongodb';

interface DocumentSchema {
  _id?: ObjectId;
}

export interface ShortIdPluginProps {
  _sid: string;
}

interface DocumentWithPluginProps extends DocumentSchema, ShortIdPluginProps {}

export interface ShortIdPluginOptions {
  sidKey?: string;
}

export default function shortIdPlugin<TSchema extends DocumentWithPluginProps>(
  model: Model<TSchema>,
  {sidKey = '_sid'}: ShortIdPluginOptions = {}
) {
  model.addSchemaProperties({
    [sidKey]: {bsonType: 'string', minLength: 7, maxLength: 14, index: {unique: true}}
  });
  model.addStatics({
    findBySid: async (sid: string, options?: FindOneOptions): Promise<TSchema | null> => {
      return model.findOne({[sidKey]: sid}, options);
    },
    updateBySid: async (
      sid: string,
      update: UpdateQuery<TSchema> | TSchema,
      options: ReplaceOneOptions = {}
    ): Promise<UpdateWriteOpResult> => {
      return model.updateOne({[sidKey]: sid}, update, options);
    },
    lazyIdFromSid: memoize(async (sid: string) => {
      const doc = await model.findOne({[sidKey]: sid}, {projection: {_id: 1}});
      return doc ? doc._id : null;
    })
  });
  model.pre('insert', (insert: TSchema) => {
    insert[sidKey] = shortid.generate();
  });
}
