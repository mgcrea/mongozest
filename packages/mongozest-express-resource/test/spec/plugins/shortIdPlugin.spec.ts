import {makeFetch} from 'supertest-fetch';
import createMongo, {shortIdPlugin as modelShortIdPlugin, ObjectId, Model} from '@mgcrea/mongozest';
import {omit, isObject} from 'lodash';
import {createTestApp, breakdownMiddleware} from './../../utils/app';
import createResource, {Resource} from './../../../src';
import shortIdPlugin from './../../../src/plugins/shortIdPlugin';

const DB_NAME = getDbName(__filename);

const app = createTestApp({routers: []});
const {mongo, insertFixture} = app.locals;
const fetch = makeFetch(app);

class User extends Model {
  static schema = {
    firstName: {bsonType: 'string'},
    lastName: {bsonType: 'string'},
    email: {bsonType: 'string', required: true},
    nationality: {bsonType: 'string'}
  };
  static plugins = [modelShortIdPlugin];
}

beforeAll(async () => {
  const db = await mongo.connect(DB_NAME);
  await db.dropDatabase();
  await mongo.loadModel(User);
});

afterAll(async () => {
  await mongo.disconnect();
});

describe('Resource', () => {
  let resource: Resource;
  describe('resource', () => {
    it('should properly create resource', async () => {
      resource = createResource('User', {db: 'mongo', plugins: [shortIdPlugin]});
      expect(resource instanceof Resource).toBeTruthy();
    });
    it('should properly build and serve resource', async () => {
      const router = resource.build();
      app.use(router);
      // @NOTE Rebind breakdown middleware to ease testing
      app.use(breakdownMiddleware);
    });
  });
  describe('document', () => {
    describe('GET /users/:_sid', () => {
      it('should return 200', async () => {
        const {ops} = await insertFixture('User');
        const {_sid} = ops[0];
        const res = await fetch(`/users/${_sid}`, {
          method: 'get',
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(isObject(resBody)).toBeTruthy();
        expect(Object.keys(resBody)).toMatchSnapshot();
        expect(omit(resBody, '_id', '_sid')).toMatchSnapshot();
      });
    });
    describe('PATCH /users/:_sid', () => {
      it('should return 200', async () => {
        const {ops} = await insertFixture('User');
        const {_sid} = ops[0];
        const reqBody = {
          firstName: 'Laura'
        };
        const res = await fetch(`/users/${_sid}`, {
          method: 'patch',
          body: JSON.stringify(reqBody),
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(isObject(resBody)).toBeTruthy();
        expect(Object.keys(resBody)).toMatchSnapshot();
        expect(omit(resBody, '_id', '_sid')).toMatchSnapshot();
      });
    });
    describe('DELETE /users/:_sid', () => {
      it('should return 200', async () => {
        const {ops} = await insertFixture('User');
        const {_sid} = ops[0];
        const res = await fetch(`/users/${_sid}`, {
          method: 'delete',
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(isObject(resBody)).toBeTruthy();
        expect(Object.keys(resBody)).toMatchSnapshot();
        expect(resBody).toMatchSnapshot();
      });
    });
  });
});
