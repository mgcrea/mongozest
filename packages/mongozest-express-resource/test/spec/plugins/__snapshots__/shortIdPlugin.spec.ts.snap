// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource document DELETE /users/:_sid should return 200 1`] = `
Array [
  "ok",
  "n",
]
`;

exports[`Resource document DELETE /users/:_sid should return 200 2`] = `
Object {
  "n": 1,
  "ok": 1,
}
`;

exports[`Resource document GET /users/:_sid should return 200 1`] = `
Array [
  "_id",
  "firstName",
  "lastName",
  "email",
  "_sid",
]
`;

exports[`Resource document GET /users/:_sid should return 200 2`] = `
Object {
  "email": "john.doe@gmail.com",
  "firstName": "John",
  "lastName": "Doe",
}
`;

exports[`Resource document PATCH /users/:_sid should return 200 1`] = `
Array [
  "_id",
  "firstName",
  "lastName",
  "email",
  "_sid",
]
`;

exports[`Resource document PATCH /users/:_sid should return 200 2`] = `
Object {
  "email": "john.doe@gmail.com",
  "firstName": "Laura",
  "lastName": "Doe",
}
`;
